File: package/main/src/Validate/isDeepEqual.ts
/**
 * Options for isDeepEqual comparison
 */
export interface IsDeepEqualOptions {
  /**
   * Whether to ignore array order when comparing arrays
   * @default true
   */
  strictOrder?: boolean;
}

/**
 * Performs a deep equality comparison between two values
 *
 * @param a - First value to compare
 * @param b - Second value to compare
 * @param options - Comparison options
 * @returns true if values are deeply equal, false otherwise
 *
 * @example
 * ```typescript
 * isDeepEqual({ a: 1, b: [2, 3] }, { b: [2, 3], a: 1 }); // true
 * isDeepEqual([1, 2, 3], [3, 2, 1]); // false
 * isDeepEqual([1, 2, 3], [3, 2, 1], { strictOrder: false }); // true
 * isDeepEqual(new Set([1, 2]), new Set([2, 1])); // true
 * isDeepEqual(new Map([['a', 1]]), new Map([['a', 1]])); // true
 * ```
 */
export function isDeepEqual(
  a: unknown,
  b: unknown,
  options: IsDeepEqualOptions = {},
): boolean {
  const { strictOrder = true } = options;
  const visited = new WeakSet<object>();

  function compare(x: unknown, y: unknown): boolean {
    if (Object.is(x, y)) {
      return true;
    }

    if (x == null || y == null) {
      return false;
    }

    if (typeof x !== typeof y) {
      return false;
    }

    if (typeof x !== "object" || typeof y !== "object") {
      return false;
    }

    if (visited.has(x) || visited.has(y)) {
      return true;
    }
    visited.add(x);
    visited.add(y);

    const ctorX = x.constructor;
    const ctorY = y.constructor;
    if (ctorX !== ctorY) {
      return false;
    }

    if (x instanceof Date && y instanceof Date) {
      return x.getTime() === y.getTime();
    }

    if (x instanceof RegExp && y instanceof RegExp) {
      return x.toString() === y.toString();
    }

    if (Array.isArray(x) && Array.isArray(y)) {
      if (x.length !== y.length) {
        return false;
      }

      if (strictOrder) {
        for (const [index, element] of x.entries()) {
          if (!compare(element, y[index])) {
            return false;
          }
        }
      } else {
        const yCopy = [...y];
        for (const itemX of x) {
          let found = false;
          for (let index = 0; index < yCopy.length; index++) {
            if (compare(itemX, yCopy[index])) {
              yCopy.splice(index, 1);
              found = true;
              break;
            }
          }
          if (!found) {
            return false;
          }
        }
        return yCopy.length === 0;
      }
      return true;
    }

    if (x instanceof Set && y instanceof Set) {
      if (x.size !== y.size) {
        return false;
      }

      for (const item of x) {
        let found = false;
        for (const otherItem of y) {
          if (compare(item, otherItem)) {
            found = true;
            break;
          }
        }
        if (!found) {
          return false;
        }
      }
      return true;
    }

    if (x instanceof Map && y instanceof Map) {
      if (x.size !== y.size) {
        return false;
      }

      for (const [key, value] of x) {
        let found = false;
        for (const [otherKey, otherValue] of y) {
          if (compare(key, otherKey) && compare(value, otherValue)) {
            found = true;
            break;
          }
        }
        if (!found) {
          return false;
        }
      }
      return true;
    }

    if (ArrayBuffer.isView(x) && ArrayBuffer.isView(y)) {
      const xArray = x as Uint8Array;
      const yArray = y as Uint8Array;
      if (xArray.byteLength !== yArray.byteLength) {
        return false;
      }
      for (let index = 0; index < xArray.byteLength; index++) {
        if (xArray[index] !== yArray[index]) {
          return false;
        }
      }
      return true;
    }

    const keysX = Object.keys(x);
    const keysY = Object.keys(y);

    if (keysX.length !== keysY.length) {
      return false;
    }

    for (const key of keysX) {
      if (!(key in y)) {
        return false;
      }
    }

    for (const key of keysX) {
      if (
        !compare(
          (x as Record<string, unknown>)[key],
          (y as Record<string, unknown>)[key],
        )
      ) {
        return false;
      }
    }

    return true;
  }

  return compare(a, b);
}

File: package/umt_python/src/validate/is_deep_equal.py
import math
from collections import Counter
from dataclasses import dataclass


@dataclass
class IsDeepEqualOptions:
    """
    Options for is_deep_equal comparison.

    Attributes:
        strict_order: Whether to ignore array order when comparing arrays.
    """

    strict_order: bool = True


def is_deep_equal(
    a: object, b: object, options: IsDeepEqualOptions | None = None
) -> bool:
    """
    Performs a deep equality comparison between two values.

    Args:
        a: First value to compare.
        b: Second value to compare.
        options: Comparison options.

    Returns:
        True if values are deeply equal, False otherwise.

    Example:
        >>> is_deep_equal({"a": 1, "b": [2, 3]}, {"b": [2, 3], "a": 1})
        True
        >>> is_deep_equal([1, 2, 3], [3, 2, 1])
        False
        >>> is_deep_equal([1, 2, 3], [3, 2, 1], IsDeepEqualOptions(strict_order=False))
        True
        >>> is_deep_equal({1, 2}, {2, 1})
        True
    """
    if options is None:
        options = IsDeepEqualOptions()
    strict_order = options.strict_order
    visited: set[int] = set()

    def compare(x: object, y: object) -> bool:
        if x is y:
            return True

        if x is None or y is None:
            return x is y

        if type(x) is not type(y):
            return False

        if isinstance(x, (int, float, str, bool)):
            if (
                isinstance(x, float)
                and isinstance(y, float)
                and math.isnan(x)
                and math.isnan(y)
            ):
                return True
            return x == y

        id_x = id(x)
        id_y = id(y)
        if id_x in visited or id_y in visited:
            return True
        visited.add(id_x)
        visited.add(id_y)

        if isinstance(x, list) and isinstance(y, list):
            if len(x) != len(y):
                return False

            if strict_order:
                for i, elem in enumerate(x):
                    if not compare(elem, y[i]):
                        return False
            else:
                try:

                    def _make_hashable_key(item: object) -> tuple[type, object]:
                        t = type(item)
                        if t is float and math.isnan(item):  # type: ignore
                            return (float, "NaN")
                        return (t, item)

                    return Counter(_make_hashable_key(i) for i in x) == Counter(
                        _make_hashable_key(i) for i in y
                    )
                except TypeError:
                    pass

                y_copy = list(y)
                for item_x in x:
                    found = False
                    for i, item_y in enumerate(y_copy):
                        if compare(item_x, item_y):
                            y_copy.pop(i)
                            found = True
                            break
                    if not found:
                        return False
                return len(y_copy) == 0
            return True

        if isinstance(x, set) and isinstance(y, set):
            if len(x) != len(y):
                return False

            for item in x:
                found = False
                for other_item in y:
                    if compare(item, other_item):
                        found = True
                        break
                if not found:
                    return False
            return True

        if isinstance(x, dict) and isinstance(y, dict):
            if len(x) != len(y):
                return False

            for key in x:
                if key not in y:
                    return False

            return all(compare(x[key], y[key]) for key in x)

        if isinstance(x, (bytes, bytearray)) and isinstance(y, (bytes, bytearray)):
            return x == y

        return x == y

    return compare(a, b)
